src/Functions/toStartOfInterval.cpp

```diff
@@ -40,156 +47,243 @@ class FunctionToStartOfInterval : public IFunction
+            const DataTypePtr & type_arg2 = arguments[1].type;
+            interval_type = checkAndGetDataType<DataTypeInterval>(type_arg2.get());
             if (!interval_type)
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
-                    "Should be an interval of time", arguments[1].type->getName(), getName());
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of 2nd argument of function {}. "
+                    "Should be an interval of time", type_arg2->getName(), getName());
             switch (interval_type->getKind()) // NOLINT(bugprone-switch-missing-default-case)
             {
                 case IntervalKind::Nanosecond:
                 case IntervalKind::Microsecond:
                 case IntervalKind::Millisecond:
-                    result_type_is_datetime_64 = true;
+                    result_type = ResultType::DateTime64;
                     break;
                 case IntervalKind::Second:
                 case IntervalKind::Minute:
                 case IntervalKind::Hour:
                 case IntervalKind::Day:
-                    result_type_is_datetime = true;
+                    result_type = ResultType::DateTime;
                     break;
                 case IntervalKind::Week:
                 case IntervalKind::Month:
                 case IntervalKind::Quarter:
                 case IntervalKind::Year:
-                    result_type_is_date = true;
+                    result_type = ResultType::Date;
                     break;
             }
         };
 
-        auto check_timezone_argument = [&]
+        enum class ThirdArgument
+        {
+            IsTimezone,
+            IsOrigin
+        };
+        ThirdArgument third_argument; /// valid only if 3rd argument is given
+        auto check_third_argument = [&]
         {
-            if (!isString(arguments[2].type))
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
-                    "This argument is optional and must be a constant string with timezone name",
+            const DataTypePtr & type_arg3 = arguments[2].type;
+            if (isString(type_arg3))
+            {
+                third_argument = ThirdArgument::IsTimezone;
+                if (value_is_date && result_type == ResultType::Date)
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                        "The timezone argument of function {} with interval type {} is allowed only when the 1st argument has the type DateTime or DateTime64",
+                        getName(), interval_type->getKind().toString());
+            }
+            else if (isDateOrDate32OrDateTimeOrDateTime64(type_arg3))
+            {
+                third_argument = ThirdArgument::IsOrigin;
+                if (isDateTime64(arguments[0].type) && isDateTime64(arguments[2].type))
```

Comment:
This does not implement what I mentioned in our discussion on Dec 5, i.e. "If we want to be correct, the return type must be the more precise type of the origin argument and the interval argument, assuming Date < DateTime < DateTime64 in terms of precision.".

But anyways, thinking about it, your code makes more sense than what I had in mind.